File: strands-ts/src/storage/in-memory-storage.ts
import type { Storage } from './storage.js'

import { normalizeKey, normalizePrefix } from './normalize.js'

/** Configuration for {@link InMemoryStorage}. */
export interface InMemoryStorageConfig {
  /**
   * Maximum number of entries before LRU eviction kicks in.
   * When a `put` would exceed this limit, the least-recently-accessed entry is evicted.
   * Set to `null` for unbounded growth.
   */
  maxEntries: number | null
}

/**
 * In-memory {@link Storage} backend backed by a `Map`.
 *
 * Useful for testing and for serverless environments where disk access is unavailable.
 * Content does not survive process restarts — for persistence use {@link LocalFileStorage}
 * or {@link S3Storage}.
 *
 * Eviction is LRU-based: when a `put` would exceed `maxEntries`, the least-recently-accessed
 * entry is evicted. This is the only eviction mechanism — there is no turn-based or
 * time-based expiration. When used with the `ContextOffloader` plugin, set `maxEntries`
 * to control how many offloaded entries are retained (each offloaded content block uses
 * exactly one key).
 *
 * Keys are normalized identically to {@link LocalFileStorage}: slash runs are collapsed,
 * leading/trailing slashes are stripped, and `..` segments are rejected.
 *
 * @example
 * ```typescript
 * const storage = new InMemoryStorage({ maxEntries: 100 })
 * await storage.put('memory/notes.json', new TextEncoder().encode('[]'))
 * const bytes = await storage.get('memory/notes.json')
 * ```
 */
export class InMemoryStorage implements Storage {
  private readonly _store = new Map<string, Uint8Array>()
  private readonly _maxEntries: number | null

  constructor(config: InMemoryStorageConfig) {
    const maxEntries = config.maxEntries
    if (maxEntries !== null && (!Number.isInteger(maxEntries) || maxEntries < 1)) {
      throw new Error('maxEntries must be a positive integer')
    }
    this._maxEntries = maxEntries
  }

  /**
   * Stores `data` under `key`, overwriting any existing value.
   * Bytes are copied on write to prevent aliasing with the caller's buffer.
   * If `maxEntries` is set and the store is full, the least-recently-accessed entry is evicted.
   *
   * @param key - Opaque, `/`-separated key identifying the value
   * @param data - Raw bytes to persist
   * @throws {@link StorageError} if the key is empty or contains `..` segments
   */
  async put(key: string, data: Uint8Array): Promise<void> {
    const normalized = normalizeKey(key)
    this._store.delete(normalized)
    if (this._maxEntries !== null && this._store.size >= this._maxEntries) {
      const oldest = this._store.keys().next().value!
      this._store.delete(oldest)
    }
    this._store.set(normalized, data.slice())
  }

  /**
   * Retrieves the bytes previously stored under `key`.
   * Returns a copy to prevent aliasing with the internal buffer.
   * Accessing a key moves it to the most-recently-used position.
   *
   * @param key - The key to read
   * @returns The stored bytes, or `null` if no value exists for `key`
   * @throws {@link StorageError} if the key is empty or contains `..` segments
   */
  async get(key: string): Promise<Uint8Array | null> {
    const normalized = normalizeKey(key)
    const value = this._store.get(normalized)
    if (value === undefined) return null
    this._store.delete(normalized)
    this._store.set(normalized, value)
    return value.slice()
  }

  /**
   * Deletes the value stored under `key`. A no-op if the key does not exist.
   *
   * @param key - The key to delete
   * @throws {@link StorageError} if the key is empty or contains `..` segments
   */
  async delete(key: string): Promise<void> {
    this._store.delete(normalizeKey(key))
  }

  /**
   * Lists the keys whose names begin with `prefix`, sorted lexicographically.
   *
   * @param prefix - Key prefix to match. An empty string matches all keys.
   * @returns The matching keys, sorted ascending
   * @throws {@link StorageError} if the prefix contains `..` segments
   */
  async list(prefix: string): Promise<string[]> {
    const normalized = normalizePrefix(prefix)
    const keys: string[] = []
    for (const key of this._store.keys()) {
      if (key.startsWith(normalized)) keys.push(key)
    }
    return keys.sort()
  }

  /**
   * Removes all stored entries. Useful for resetting state between tests.
   */
  clear(): void {
    this._store.clear()
  }
}

File: strands-py/src/strands/storage/in_memory_storage.py
"""In-memory storage implementation."""

from __future__ import annotations

import builtins
import threading
from typing import TYPE_CHECKING

from .search.keyword import KeywordSearchStrategy
from .storage import _NamespacedStorage, _normalize_key, _normalize_prefix

if TYPE_CHECKING:
    from .search.types import SearchStrategy
    from .storage import StorageSearchResult


class InMemoryStorage:
    """Map-backed storage for testing and short-lived processes.

    Content does not survive process restarts. The store is unbounded — consumers
    manage eviction themselves.

    Example:
        ```python
        from strands.storage import InMemoryStorage

        storage = InMemoryStorage()
        await storage.write("sessions/abc/state.json", b'{"messages": []}')
        data = await storage.read("sessions/abc/state.json")
        ```
    """

    def __init__(self, *, search_strategy: SearchStrategy[InMemoryStorage] | None = None) -> None:
        """Initialize an empty in-memory store.

        Args:
            search_strategy: Optional search strategy. When set, ``write()``
                automatically indexes entries and ``search()`` delegates to the
                strategy instead of the default keyword scan.
        """
        self._store: dict[str, bytes] = {}
        self._lock = threading.Lock()
        self._search_strategy = search_strategy

    async def write(self, key: str, data: bytes) -> None:
        """Store data under key, overwriting any existing value.

        Args:
            key: Opaque string key identifying the value.
            data: Raw bytes to persist.

        Raises:
            StorageError: If the key is invalid.
        """
        normalized = _normalize_key(key)
        with self._lock:
            self._store[normalized] = bytes(data)

        if self._search_strategy is not None:
            try:
                await self._search_strategy.index(self, normalized, data)
            except Exception as error:
                from ..types.exceptions import StorageError

                raise StorageError(f"Wrote '{key}' but indexing failed") from error

    async def read(self, key: str) -> bytes | None:
        """Retrieve the bytes previously stored under key.

        Args:
            key: The key to read.

        Returns:
            The stored bytes, or None if no value exists for key.

        Raises:
            StorageError: If the key is invalid.
        """
        normalized = _normalize_key(key)
        with self._lock:
            value = self._store.get(normalized)
        return value

    async def delete(self, key: str) -> None:
        """Delete the value stored under key. A no-op if the key does not exist.

        Args:
            key: The key to delete.

        Raises:
            StorageError: If the key is invalid.
        """
        normalized = _normalize_key(key)
        with self._lock:
            self._store.pop(normalized, None)

    async def list(self, query: str = "") -> builtins.list[str]:
        """List keys matching the given prefix.

        Args:
            query: A prefix string to filter keys. Empty string matches all.

        Returns:
            Matching keys sorted ascending.

        Raises:
            StorageError: If the prefix is invalid.
        """
        prefix = _normalize_prefix(query)
        with self._lock:
            keys = sorted(k for k in self._store if k.startswith(prefix))
        return keys

    async def search(self, query: str) -> builtins.list[StorageSearchResult]:
        """Search stored content using the configured strategy.

        Delegates to the search strategy when one is set, otherwise falls back
        to keyword token-overlap scoring.

        Args:
            query: Natural-language search query.

        Returns:
            All matches with relevance scores, ranked best-first.
        """
        if self._search_strategy is not None:
            return await self._search_strategy.search(self, query)
        return await KeywordSearchStrategy().search(self, query)

    def namespace(self, prefix: str) -> _NamespacedStorage:
        """Return a view of this storage with all keys prefixed.

        Args:
            prefix: Prefix to prepend to all keys.

        Returns:
            A namespaced storage view.
        """
        return _NamespacedStorage(self, prefix)

    def clear(self) -> None:
        """Remove all stored entries."""
        with self._lock:
            self._store.clear()
